apps/opt.c

```diff
@@ -656,22 +659,54 @@ int opt_next(void)
             /* Just a string. */
             break;
         case '/':
-            if (app_isdir(arg) >= 0)
+            if (app_isdir(arg) > 0)
                 break;
             BIO_printf(bio_err, "%s: Not a directory: %s\n", prog, arg);
             return -1;
         case '<':
             /* Input file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) >= 0)
+            if (strcmp(arg, "-") == 0 || app_access(arg, R_OK) == 0)
                 break;
             BIO_printf(bio_err,
                        "%s: Cannot open input file %s, %s\n",
                        prog, arg, strerror(errno));
             return -1;
         case '>':
             /* Output file. */
-            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) >= 0 || errno == ENOENT)
+            l = strlen(arg);
+            if (arg[l - 1] == '/' || app_isdir(arg) > 0) {
+                BIO_printf(bio_err, "%s: Is a directory: %s\n", prog, arg);
+                return -1;
+            }
+            if (strcmp(arg, "-") == 0 || app_access(arg, W_OK) == 0
+                || errno == ENOENT) {
+#if !defined(_WIN32)
+                char *c = NULL, *d = NULL;
+
+                if (errno == ENOENT) {
```

Comment:
Is there guarantee that errno is not ENOENT if other condition is met. For example if you ended up here because of strcmp(arg, "-")...